// @openserv-labs/sdk/dist/mcp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MCPErrorCodes = exports.McpError = exports.MCPClient = exports.mcpServerConfigSchema = exports.mcpServerConfigHttpSchema = exports.mcpServerConfigSseSchema = exports.mcpServerConfigStdioSchema = void 0;
exports.importClient = importClient;
exports.importStdioTransport = importStdioTransport;
exports.importSSETransport = importSSETransport;
exports.importHttpTransport = importHttpTransport;
exports.createClient = createClient;
exports.createStdioTransport = createStdioTransport;
exports.createSSETransport = createSSETransport;
exports.createHttpTransport = createHttpTransport;
const zod_1 = require("zod");
const logger_1 = require("./logger");
// Zod Schemas for MCP Server Configuration
/** Zod schema for validating MCP server configurations using 'stdio' transport. */
exports.mcpServerConfigStdioSchema = zod_1.z.object({
    /** Specifies the transport type as 'stdio'. */
    transport: zod_1.z.literal('stdio'),
    /** The command to execute to start the server. */
    command: zod_1.z.string().min(1, 'Command cannot be empty'),
    /** Arguments to pass to the command. */
    args: zod_1.z.array(zod_1.z.string()),
    /** Optional environment variables for the command. */
    env: zod_1.z.record(zod_1.z.string()).optional(),
    /** Whether to automatically discover tools and register them as capabilities upon successful connection. Defaults to false. */
    autoRegisterTools: zod_1.z.boolean().optional()
});
/** Zod schema for validating MCP server configurations using 'sse' transport. */
exports.mcpServerConfigSseSchema = zod_1.z.object({
    /** Specifies the transport type as 'sse'. */
    transport: zod_1.z.literal('sse'),
    /** The URL of the MCP server's SSE endpoint. */
    url: zod_1.z.string().url('Invalid URL format for SSE transport'),
    /** Whether to automatically discover tools and register them as capabilities upon successful connection. Defaults to false. */
    autoRegisterTools: zod_1.z.boolean().optional()
});
/** Zod schema for validating MCP server configurations using 'http' (Streamable HTTP) transport. */
exports.mcpServerConfigHttpSchema = zod_1.z.object({
    /** Specifies the transport type as 'http'. */
    transport: zod_1.z.literal('http'),
    /** The URL of the MCP server's HTTP endpoint. */
    url: zod_1.z.string().url('Invalid URL format for HTTP transport'),
    /** Whether to automatically discover tools and register them as capabilities upon successful connection. Defaults to false. */
    autoRegisterTools: zod_1.z.boolean().optional()
});
/**
 * Discriminated union Zod schema for validating MCP server configurations.
 * It differentiates based on the 'transport' property.
 */
exports.mcpServerConfigSchema = zod_1.z.discriminatedUnion('transport', [
    exports.mcpServerConfigStdioSchema,
    exports.mcpServerConfigSseSchema,
    exports.mcpServerConfigHttpSchema
]);
/** Dynamically imports the MCP Client class from the MCP SDK. */
async function importClient() {
    const { Client } = await import('@modelcontextprotocol/sdk/client/index.js');
    return Client;
}
/** Dynamically imports the MCP StdioClientTransport class from the MCP SDK. */
async function importStdioTransport() {
    const { StdioClientTransport } = await import('@modelcontextprotocol/sdk/client/stdio.js');
    return StdioClientTransport;
}
/** Dynamically imports the MCP SSEClientTransport class from the MCP SDK. */
async function importSSETransport() {
    const { SSEClientTransport } = await import('@modelcontextprotocol/sdk/client/sse.js');
    return SSEClientTransport;
}
/** Dynamically imports the MCP StreamableHTTPClientTransport class from the MCP SDK. */
async function importHttpTransport() {
    const { StreamableHTTPClientTransport } = await import('@modelcontextprotocol/sdk/client/streamableHttp.js');
    return StreamableHTTPClientTransport;
}
/**
 * Creates a new MCP client instance.
 * @param name - The name of the client.
 * @param version - The version of the client.
 * @returns A promise that resolves to an MCPClient instance.
 */
async function createClient(name, version) {
    const Client = await importClient();
    return new Client({ name, version });
}
/**
 * Creates a new MCP transport layer for stdio communication.
 * @param options - Configuration options for the stdio transport.
 * @param options.command - The command to execute to start the server.
 * @param options.args - Arguments to pass to the command.
 * @param options.env - Optional environment variables for the command.
 * @returns A promise that resolves to an MCPTransport instance for stdio.
 */
async function createStdioTransport(options) {
    const StdioClientTransport = await importStdioTransport();
    return new StdioClientTransport(options);
}
/**
 * Creates a new MCP transport layer for Server-Sent Events (SSE) communication.
 * @param url - The URL of the MCP server's SSE endpoint.
 * @returns A promise that resolves to an MCPTransport instance for SSE.
 */
async function createSSETransport(url) {
    const SSEClientTransport = await importSSETransport();
    return new SSEClientTransport(url);
}
/**
 * Creates a new MCP transport layer for Streamable HTTP communication.
 * @param url - The URL of the MCP server's HTTP endpoint.
 * @returns A promise that resolves to an MCPTransport instance for HTTP.
 */
async function createHttpTransport(url) {
    const StreamableHTTPClientTransport = await importHttpTransport();
    return new StreamableHTTPClientTransport(url);
}
/**
 * MCP server class for handling MCP protocol communication.
 * Manages tool registration and execution.
 */
class MCPClient {
    /** The unique identifier for this MCP server configuration. */
    serverId;
    /** Array storing discovered tool descriptors after connection. */
    tools = [];
    /** The configuration object for this specific MCP server. */
    config;
    /** The underlying client instance from the @modelcontextprotocol/sdk. */
    client;
    /**
     * Creates a new MCPClient instance.
     * Validates the provided configuration.
     *
     * @param serverId - The unique identifier for this MCP server configuration.
     * @param config - The configuration object for the server.
     * @throws {McpError} If the configuration is invalid.
     */
    constructor(serverId, config) {
        this.config = config;
        this.serverId = serverId;
        if (!this.validateMCPServerConfig(this.serverId, this.config)) {
            throw new McpError(exports.MCPErrorCodes.INVALID_PARAMS, `Invalid configuration for MCP server "${this.serverId}". Check logs for details.`);
        }
    }
    /**
     * Connects to the MCP server using the configured transport.
     * Establishes a connection based on the transport type (stdio or sse)
     * and initializes the client with the appropriate transport layer.
     * @returns {Promise<void>} A promise that resolves when the connection is established and tools (if any) are fetched.
     * @throws {McpError} If connection fails or transport type is invalid
     */
    async connect() {
        let transport;
        const transportType = this.config.transport;
        try {
            logger_1.logger.info(`Connecting to MCP server "${this.serverId}" using ${transportType}...`);
            if (this.config.transport === 'sse') {
                transport = await createSSETransport(new URL(this.config.url));
            }
            else if (this.config.transport === 'http') {
                transport = await createHttpTransport(new URL(this.config.url));
            }
            else if (this.config.transport === 'stdio') {
                transport = await createStdioTransport({
                    command: this.config.command,
                    args: this.config.args,
                    env: this.config.env
                });
            }
            if (!transport) {
                throw new McpError(exports.MCPErrorCodes.INTERNAL_ERROR, `Transport could not be initialized for "${this.serverId}" with type ${transportType}.`);
            }
            this.client = await createClient(`agent-${this.serverId}`, '1.0.0');
            this.client.connect(transport);
            // Wait for the server to be ready
            await new Promise(resolve => setTimeout(resolve, 5000));
            logger_1.logger.info(`Successfully connected to MCP server "${this.serverId}"`);
            this.tools = await this.fetchMCPToolDescriptors();
        }
        catch (error) {
            if (this.client)
                this.client.disconnect();
            if (error instanceof McpError) {
                throw error;
            }
            throw new McpError(exports.MCPErrorCodes.INTERNAL_ERROR, `Failed to connect to MCP server "${this.serverId}" using ${transportType}: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    /**
     * Fetches the list of tool descriptors from the connected MCP server.
     * Should only be called after a successful connection.
     *
     * @returns {Promise<MCPToolDescriptor[]>} A promise resolving to an array of tool descriptors.
     * @private
     * @throws {McpError} If the client is not connected or the server response is invalid.
     */
    async fetchMCPToolDescriptors() {
        if (!this.client) {
            throw new McpError(exports.MCPErrorCodes.INVALID_PARAMS, `Cannot fetch tool descriptors: No active client found for server "${this.serverId}". Ensure connection was successful.`);
        }
        // Request tool list
        const response = await this.client.listTools();
        if (!response || !Array.isArray(response.tools)) {
            logger_1.logger.warn(`No tools returned from "${this.serverId}"`);
            return [];
        }
        return response.tools;
    }
    /**
     * Validates the MCP server configuration using the Zod schema.
     *
     * @param serverName - The name/ID of the server (for logging).
     * @param config - The configuration object to validate.
     * @returns `true` if valid, `false` otherwise.
     * @private
     */
    validateMCPServerConfig(serverName, config) {
        const result = exports.mcpServerConfigSchema.safeParse(config);
        if (!result.success) {
            logger_1.logger.error(`Invalid configuration for MCP server "${serverName}". Errors: `, result.error.flatten());
            return false;
        }
        return true;
    }
    /**
     * Executes a specific tool on the connected MCP server.
     *
     * @param toolName - The name of the tool to execute.
     * @param parameters - The parameters to pass to the tool.
     * @returns A promise resolving with the result from the tool execution.
     * @throws {McpError} If the client is not connected or the tool execution fails.
     */
    async executeTool(toolName, parameters) {
        if (!this.client) {
            throw new McpError(exports.MCPErrorCodes.CONNECTION_CLOSED, 'Client not connected or initialized');
        }
        try {
            const result = await this.client.callTool({
                name: toolName,
                arguments: parameters
            });
            return result;
        }
        catch (e) {
            throw new McpError(exports.MCPErrorCodes.INTERNAL_ERROR, `Failed to execute tool ${toolName} on server ${this.serverId}: ${e instanceof Error ? e.message : String(e)}`);
        }
    }
    /**
     * Gets the list of discovered tool descriptors for this server.
     * Tools are fetched during the `connect` method.
     *
     * @returns {MCPToolDescriptor[]} An array of discovered tool descriptors.
     */
    getTools() {
        return this.tools;
    }
    /**
     * Gets the underlying protocol client instance.
     * Primarily for internal use or specific low-level interactions.
     *
     * @returns {ProtocolClient} The raw protocol client instance.
     */
    getClient() {
        return this.client;
    }
}
exports.MCPClient = MCPClient;
/**
 * Standard MCP error class with error codes following JSON-RPC spec.
 * Useful for creating and identifying MCP-specific errors.
 */
class McpError extends Error {
    /** The JSON-RPC error code. */
    code;
    /** Optional additional data related to the error. */
    data;
    /**
     * Constructs an McpError.
     * @param code - The JSON-RPC error code (see MCPErrorCodes).
     * @param message - A human-readable description of the error.
     * @param data - Optional additional data related to the error.
     */
    constructor(code, message, data) {
        super(`MCP error ${code}: ${message}`);
        this.name = 'McpError';
        this.code = code;
        this.data = data;
    }
}
exports.McpError = McpError;
/**
 * MCP error codes following the JSON-RPC specification.
 * Provides a standardized set of codes for common MCP errors.
 */
exports.MCPErrorCodes = {
    /** Connection was closed. */
    CONNECTION_CLOSED: -32000,
    /** Request timed out. */
    REQUEST_TIMEOUT: -32001,
    /** Invalid JSON was received by the server. An error occurred on the server while parsing the JSON text. */
    PARSE_ERROR: -32700,
    /** The JSON sent is not a valid Request object. */
    INVALID_REQUEST: -32600,
    /** The method does not exist / is not available. */
    METHOD_NOT_FOUND: -32601,
    /** Invalid method parameter(s). */
    INVALID_PARAMS: -32602,
    /** Internal JSON-RPC error. */
    INTERNAL_ERROR: -32603
};
