// @openserv-labs/client/dist/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformClient = void 0;
const axios_1 = __importDefault(require("axios"));
const accounts_1 = require("viem/accounts");
const utils_1 = require("./utils");
const agents_api_1 = require("./agents-api");
const integrations_api_1 = require("./integrations-api");
const models_api_1 = require("./models-api");
const triggers_api_1 = require("./triggers-api");
const tasks_api_1 = require("./tasks-api");
const workflows_api_1 = require("./workflows-api");
const web3_api_1 = require("./web3-api");
const payments_api_1 = require("./payments-api");
const erc8004_api_1 = require("./erc8004-api");
const PLATFORM_URL = process.env.OPENSERV_API_URL || "https://api.openserv.ai";
/**
 * Client for interacting with the OpenServ Platform API.
 *
 * @example
 * ```typescript
 * // Using API key authentication
 * const client = new PlatformClient({ apiKey: 'your-api-key' });
 *
 * // Using environment variable
 * const client = new PlatformClient(); // Uses OPENSERV_USER_API_KEY
 *
 * // Using wallet authentication
 * const client = new PlatformClient();
 * await client.authenticate(process.env.WALLET_PRIVATE_KEY);
 * ```
 */
class PlatformClient {
    _apiClient;
    /** Wallet address, set by authenticate() or manually. Used as a fallback for x402 trigger wallet resolution. */
    walletAddress;
    /** API for managing agents */
    agents;
    /** API for managing integration connections */
    integrations;
    /** API for discovering available LLM models */
    models;
    /** API for managing workflow triggers */
    triggers;
    /** API for managing workflow tasks */
    tasks;
    /** API for managing workflows */
    workflows;
    /** API for Web3 operations (USDC top-up) */
    web3;
    /** API for x402 payments to access paid workflows */
    payments;
    /** API for ERC-8004 agent identity (deployment, wallets, IPFS, reputation) */
    erc8004;
    /**
     * Get the raw axios client for advanced use cases.
     * @returns The underlying axios instance
     */
    get rawClient() {
        return this._apiClient;
    }
    /**
     * Make a GET request to the API.
     * @param path - API endpoint path
     * @param config - Optional Axios request config
     * @returns Response data
     */
    async get(path, config) {
        const response = await this._apiClient.get(path, config);
        return response.data;
    }
    /**
     * Make a POST request to the API.
     * @param path - API endpoint path
     * @param data - Request body data
     * @param config - Optional Axios request config
     * @returns Response data
     */
    async post(path, data, config) {
        const response = await this._apiClient.post(path, data, config);
        return response.data;
    }
    /**
     * Make a PUT request to the API.
     * @param path - API endpoint path
     * @param data - Request body data
     * @param config - Optional Axios request config
     * @returns Response data
     */
    async put(path, data, config) {
        const response = await this._apiClient.put(path, data, config);
        return response.data;
    }
    /**
     * Make a DELETE request to the API.
     * @param path - API endpoint path
     * @param config - Optional Axios request config
     * @returns Response data
     */
    async delete(path, config) {
        const response = await this._apiClient.delete(path, config);
        return response.data;
    }
    /**
     * Creates a new PlatformClient instance.
     *
     * @param options - Configuration options
     * @param options.apiKey - API key for authentication (defaults to OPENSERV_USER_API_KEY env var)
     * @param options.baseUrl - Base URL for the API (defaults to https://api.openserv.ai)
     */
    constructor(options) {
        const apiKey = options?.apiKey || process.env.OPENSERV_USER_API_KEY;
        const baseUrl = options?.baseUrl || PLATFORM_URL;
        this._apiClient = axios_1.default.create({
            baseURL: baseUrl,
            headers: {
                "Content-Type": "application/json",
                ...(apiKey ? { "x-openserv-key": apiKey } : {}),
            },
        });
        this.agents = new agents_api_1.AgentsAPI(this);
        this.integrations = new integrations_api_1.IntegrationsAPI(this);
        this.models = new models_api_1.ModelsAPI(this);
        this.triggers = new triggers_api_1.TriggersAPI(this);
        this.tasks = new tasks_api_1.TasksAPI(this);
        this.workflows = new workflows_api_1.WorkflowsAPI(this);
        this.web3 = new web3_api_1.Web3API(this);
        this.payments = new payments_api_1.PaymentsAPI(this);
        this.erc8004 = new erc8004_api_1.Erc8004API(this);
    }
    /**
     * Authenticate using an Ethereum wallet (SIWE - EIP-4361).
     *
     * This method performs Sign-In with Ethereum authentication:
     * 1. Gets a nonce from the platform
     * 2. Signs the SIWE message with the wallet
     * 3. Verifies the signature and receives an API key
     *
     * @param privateKey - Wallet private key (defaults to WALLET_PRIVATE_KEY env var)
     * @returns The API key received from authentication
     *
     * @example
     * ```typescript
     * const client = new PlatformClient();
     * const apiKey = await client.authenticate(process.env.WALLET_PRIVATE_KEY);
     * // Client is now authenticated and ready to use
     * ```
     */
    async authenticate(privateKey) {
        const walletKey = privateKey || process.env.WALLET_PRIVATE_KEY;
        if (!walletKey) {
            // If no wallet key, assume API key auth is already set up
            return "";
        }
        // Create account and store address for x402 wallet resolution
        const account = (0, accounts_1.privateKeyToAccount)((0, utils_1.normalizePrivateKey)(walletKey));
        const walletAddress = account.address;
        this.walletAddress = walletAddress;
        // Step 1: Get nonce from platform
        const nonceResponse = await this._apiClient.post("/auth/wallet/nonce", {
            walletAddress,
        });
        const { nonce, message: legacyMessage } = nonceResponse.data;
        if (!nonce && !legacyMessage) {
            throw new Error(`Nonce endpoint did not return nonce or message. Response: ${JSON.stringify(nonceResponse.data)}`);
        }
        // Construct SIWE-formatted message (EIP-4361)
        const domain = "platform.openserv.ai";
        const uri = "https://platform.openserv.ai";
        const issuedAt = new Date().toISOString();
        const message = legacyMessage ||
            `${domain} wants you to sign in with your Ethereum account:
${walletAddress}

Sign in to OpenServ

URI: ${uri}
Version: 1
Chain ID: 1
Nonce: ${nonce}
Issued At: ${issuedAt}`;
        // Step 2: Sign the message
        const signature = await account.signMessage({ message });
        // Step 3: Verify signature and get API key
        const verifyResponse = await this._apiClient.post("/auth/wallet/verify", {
            walletAddress,
            signature,
            message,
        });
        const apiKey = verifyResponse.data.apiKey;
        // Update client with API key header
        this._apiClient.defaults.headers.common["x-openserv-key"] = apiKey;
        return apiKey;
    }
    /**
     * Resolve wallet address from stored state or environment.
     *
     * Returns `this.walletAddress` (set by `authenticate()`) if available,
     * otherwise derives the address from `WALLET_PRIVATE_KEY` environment variable.
     * Per-trigger overrides are handled at the call site, not here.
     *
     * @returns The resolved wallet address, or undefined if no wallet is available
     */
    resolveWalletAddress() {
        if (this.walletAddress)
            return this.walletAddress;
        if (process.env.WALLET_PRIVATE_KEY) {
            return (0, accounts_1.privateKeyToAccount)((0, utils_1.normalizePrivateKey)(process.env.WALLET_PRIVATE_KEY)).address;
        }
        return undefined;
    }
}
exports.PlatformClient = PlatformClient;
