// @openserv-labs/client/dist/integrations-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntegrationsAPI = void 0;
class IntegrationsAPI {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * Get list of integration connections for the authenticated user
     */
    async listConnections() {
        return this.client.get("/integration/connections");
    }
    /**
     * Create an integration connection for a custom integration (like manual-trigger, webhook-trigger, etc.)
     */
    async connect(params) {
        return this.client.post("/integration/connection", {
            type: "custom",
            identifier: params.identifier,
            props: params.props || {},
        });
    }
    /**
     * Get the integration connection ID for a specific integration identifier.
     * If no connection exists, creates one for custom integrations.
     */
    async getOrCreateConnection(identifier) {
        const connections = await this.listConnections();
        const existing = connections.find((c) => c.integrationName === identifier);
        if (existing) {
            return existing.id;
        }
        // Create a new connection for custom integrations
        await this.connect({ identifier });
        // Fetch again to get the new connection ID
        const updatedConnections = await this.listConnections();
        const newConnection = updatedConnections.find((c) => c.integrationName === identifier);
        if (!newConnection) {
            throw new Error(`Failed to create integration connection for ${identifier}`);
        }
        return newConnection.id;
    }
}
exports.IntegrationsAPI = IntegrationsAPI;
