// @openserv-labs/client/dist/tasks-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TasksAPI = void 0;
/**
 * API for managing tasks within workflows.
 *
 * @example
 * ```typescript
 * const client = new PlatformClient({ apiKey: 'your-key' });
 *
 * // Create a task
 * const task = await client.tasks.create({
 *   workflowId: 123,
 *   agentId: 456,
 *   description: 'Process the data',
 *   body: 'Additional details'
 * });
 *
 * // Create a task with dependencies
 * const task2 = await client.tasks.create({
 *   workflowId: 123,
 *   agentId: 456,
 *   description: 'Follow-up task',
 *   dependencies: [task.id]
 * });
 * ```
 */
class TasksAPI {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * Create a new task in a workflow.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID to create the task in
     * @param params.agentId - The agent ID to assign the task to
     * @param params.description - Short description of the task
     * @param params.body - Detailed task body (defaults to description)
     * @param params.input - Input data for the task
     * @param params.dependencies - Array of task IDs this task depends on
     * @returns The created task
     */
    async create(params) {
        const { workflowId, agentId, description, body, input, dependencies } = params;
        const data = await this.client.post(`/workspaces/${workflowId}/task`, {
            assignee: typeof agentId === "string" ? parseInt(agentId) : agentId,
            description: description,
            body: body || description,
            input: input || "",
            outputOptions: {
                default: {
                    name: "Task Output",
                    type: "text",
                    instructions: "Complete the task and provide output",
                },
            },
            dependencies: dependencies || [],
        });
        return this.get({ workflowId, id: data.id });
    }
    /**
     * Get a task by ID.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The task ID
     * @returns The task
     */
    async get(params) {
        return this.client.get(`/workspaces/${params.workflowId}/tasks/${params.id}`);
    }
    /**
     * List all tasks in a workflow.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @returns Array of tasks
     */
    async list(params) {
        return this.client.get(`/workspaces/${params.workflowId}/tasks`);
    }
    /**
     * Update an existing task.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The task ID to update
     * @param params.description - New description (optional)
     * @param params.body - New body (optional)
     * @param params.input - New input (optional)
     * @param params.status - New status (optional)
     * @param params.assigneeAgentId - New assignee agent ID (optional)
     * @param params.dependencies - New dependencies (optional)
     * @returns The updated task
     */
    async update(params) {
        const { workflowId, id, ...data } = params;
        await this.client.put(`/workspaces/${workflowId}/tasks/${id}`, data);
        // PUT returns success message for user auth, fetch the updated task
        return this.get({ workflowId, id });
    }
    /**
     * Delete a task.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The task ID to delete
     */
    async delete(params) {
        await this.client.delete(`/workspaces/${params.workflowId}/tasks/${params.id}`);
    }
}
exports.TasksAPI = TasksAPI;
