// @openserv-labs/client/dist/triggers-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggersAPI = exports.triggers = void 0;
exports.inputSchemaToJsonSchema = inputSchemaToJsonSchema;
// ============================================================================
// Trigger Factory Functions
// ============================================================================
/**
 * Factory functions for creating type-safe trigger configurations.
 * Each factory accepts user-friendly parameters and returns a flat,
 * properly typed config with API-ready field names.
 *
 * @example
 * ```typescript
 * import { triggers } from '@openserv-labs/client';
 *
 * const webhook = triggers.webhook({ waitForCompletion: true });
 * // { type: 'webhook', waitForCompletion: true, timeout: 600 }
 *
 * const x402 = triggers.x402({ price: '0.01' });
 * // { type: 'x402', x402Pricing: '0.01', timeout: 600 }
 * ```
 */
exports.triggers = {
    /**
     * Create a webhook trigger configuration.
     * @param opts - Options for the webhook trigger
     * @returns Webhook trigger configuration
     */
    webhook: (opts) => ({
        type: "webhook",
        ...(opts?.name && { name: opts.name }),
        ...(opts?.description && { description: opts.description }),
        waitForCompletion: opts?.waitForCompletion ?? false,
        timeout: opts?.timeout ?? 600,
        ...(opts?.input && {
            inputSchema: inputSchemaToJsonSchema(opts.input),
        }),
    }),
    /**
     * Create an x402 (paid) trigger configuration.
     * @param opts - Options for the x402 trigger
     * @param opts.name - Display name for the service (e.g., "AI Research Assistant")
     * @param opts.description - Description of what this service does
     * @param opts.price - Price in USD (e.g., "0.01")
     * @returns x402 trigger configuration
     */
    x402: (opts) => ({
        type: "x402",
        ...(opts.name && { name: opts.name }),
        ...(opts.description && { description: opts.description }),
        x402Pricing: opts.price,
        waitForCompletion: true,
        timeout: opts.timeout ?? 600,
        ...(opts.walletAddress && {
            x402WalletAddress: opts.walletAddress,
        }),
        ...(opts.input && {
            inputSchema: inputSchemaToJsonSchema(opts.input),
        }),
    }),
    /**
     * Create a cron (scheduled) trigger configuration.
     * @param opts - Options for the cron trigger
     * @param opts.schedule - Cron expression
     * @param opts.timezone - Timezone (default: "UTC")
     * @returns Cron trigger configuration
     */
    cron: (opts) => ({
        type: "cron",
        ...(opts.name && { name: opts.name }),
        ...(opts.description && { description: opts.description }),
        schedule: opts.schedule,
        timezone: opts.timezone || "UTC",
    }),
    /**
     * Create a manual trigger configuration.
     * @param opts - Optional settings for the manual trigger
     * @returns Manual trigger configuration
     */
    manual: (opts) => ({
        type: "manual",
        ...(opts?.name && { name: opts.name }),
        ...(opts?.description && { description: opts.description }),
    }),
};
// ============================================================================
// Helpers
// ============================================================================
/**
 * Convert an InputSchema to JSON Schema format.
 * @param input - The input schema to convert
 * @returns JSON Schema compliant object
 */
function inputSchemaToJsonSchema(input) {
    const properties = {};
    const required = [];
    for (const [key, prop] of Object.entries(input)) {
        properties[key] = {
            type: prop.type,
            ...(prop.title && { title: prop.title }),
            ...(prop.description && { description: prop.description }),
            ...(prop.enum && { enum: prop.enum }),
            ...(prop.default !== undefined && { default: prop.default }),
        };
        if (prop.default === undefined) {
            required.push(key);
        }
    }
    return {
        $schema: "http://json-schema.org/draft-07/schema#",
        type: "object",
        properties,
        required,
    };
}
/**
 * API for managing workflow triggers.
 *
 * @example
 * ```typescript
 * const client = new PlatformClient({ apiKey: 'your-key' });
 *
 * // Get integration connection ID
 * const connId = await client.integrations.getOrCreateConnection('webhook-trigger');
 *
 * // Create a trigger
 * const trigger = await client.triggers.create({
 *   workflowId: 123,
 *   name: 'My Webhook',
 *   integrationConnectionId: connId,
 *   props: { waitForCompletion: true }
 * });
 *
 * // Activate the trigger
 * await client.triggers.activate({ workflowId: 123, id: trigger.id });
 * ```
 */
class TriggersAPI {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * Create a new trigger in a workflow.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.name - Display name for the trigger (e.g., "AI Research Assistant")
     * @param params.description - Description of what this trigger does
     * @param params.integrationConnectionId - Integration connection ID (e.g., from getOrCreateConnection)
     * @param params.props - Trigger properties
     * @param params.trigger_name - Optional specific trigger name
     * @returns The created trigger
     */
    async create(params) {
        const { workflowId, name, description, integrationConnectionId, props, trigger_name, } = params;
        const data = await this.client.post(`/workspaces/${workflowId}/trigger`, {
            name,
            description: description || name,
            // If trigger_name not specified, server will use the integration's default
            ...(trigger_name && { trigger_name }),
            integrationConnectionId,
            props: props || {},
            attributes: {},
        });
        if (!data.id) {
            throw new Error(`Trigger creation returned no ID. Response: ${JSON.stringify(data)}`);
        }
        const triggerId = data.id;
        // The create response may include a token; capture it if present
        const createToken = data.token;
        // The single trigger GET endpoint doesn't include token,
        // so fetch from list (workspace response) which includes it
        const triggersList = await this.list({ workflowId });
        const triggerFromList = triggersList.find((t) => t.id === triggerId);
        if (triggerFromList) {
            return triggerFromList;
        }
        // Fallback to get endpoint if not found in list (shouldn't happen)
        const triggerDetails = await this.get({ workflowId, id: triggerId });
        return {
            ...triggerDetails,
            id: triggerId,
            token: createToken,
        };
    }
    /**
     * Get a trigger by ID.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The trigger ID
     * @returns The trigger
     */
    async get(params) {
        const data = await this.client.get(`/workspaces/${params.workflowId}/triggers/${params.id}`);
        // The single trigger GET endpoint doesn't return the ID in the response,
        // so we include it from the request parameters
        return {
            ...data,
            id: params.id,
        };
    }
    /**
     * List all triggers in a workflow.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @returns Array of triggers
     */
    async list(params) {
        // The list endpoint returns available integrations, not actual trigger instances
        // To get actual triggers, we need to fetch from the workspace get endpoint
        const workspace = await this.client.get(`/workspaces/${params.workflowId}`);
        // The workspace response includes triggers with the full structure
        const triggers = workspace.triggers || [];
        // Map to Trigger type - token is in attributes.uiState.token
        return triggers.map((t) => ({
            id: t.id,
            name: t.name,
            description: t.description || undefined,
            integrationConnectionId: t.integrationConnection?.id || "",
            props: t.props,
            token: t.attributes?.uiState?.token,
            isActive: t.is_active,
            state: t.state,
        }));
    }
    /**
     * Update an existing trigger.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The trigger ID to update
     * @param params.props - New properties (optional)
     * @param params.name - New name (optional)
     * @param params.description - New description (optional)
     * @returns The updated trigger
     */
    async update(params) {
        const { workflowId, id, props, name, description } = params;
        // First get the current trigger to preserve required fields
        const currentTrigger = await this.get({ workflowId, id });
        return this.client.put(`/workspaces/${workflowId}/triggers/${id}`, {
            name: name ?? currentTrigger.name,
            description: description ?? currentTrigger.description ?? currentTrigger.name,
            integrationConnectionId: currentTrigger.integrationConnectionId,
            props: props ?? currentTrigger.props,
        });
    }
    /**
     * Delete a trigger.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The trigger ID to delete
     */
    async delete(params) {
        await this.client.delete(`/workspaces/${params.workflowId}/triggers/${params.id}`);
    }
    /**
     * Activate a trigger so it can receive events.
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The trigger ID to activate
     */
    async activate(params) {
        const currentTrigger = await this.get({
            workflowId: params.workflowId,
            id: params.id,
        });
        await this.client.put(`/workspaces/${params.workflowId}/triggers/${params.id}`, {
            name: currentTrigger.name,
            integrationConnectionId: currentTrigger.integrationConnectionId,
            props: currentTrigger.props,
            is_active: true,
        });
    }
    /**
     * Fire a trigger manually (for testing or manual invocation).
     * @param params - Parameters object
     * @param params.workflowId - The workflow ID
     * @param params.id - The trigger ID to fire
     * @param params.input - Optional input data as JSON string
     * @returns Response from the trigger
     */
    async fire(params) {
        return this.client.put(`/workspaces/${params.workflowId}/triggers/${params.id}/trigger`, {
            input: params.input || "",
        });
    }
    /**
     * Fire a webhook trigger by workflow ID or direct URL.
     *
     * When `workflowId` is provided, the method automatically resolves the
     * webhook trigger token by listing the workflow's triggers and finding
     * the first non-x402 trigger with a token. You can narrow which trigger
     * to use by also passing `triggerId` or `triggerName`.
     *
     * @param params - Parameters
     * @param params.workflowId - The workflow ID (recommended - resolves webhook URL automatically)
     * @param params.triggerUrl - Direct webhook URL (alternative to workflowId)
     * @param params.triggerId - Specific trigger ID within the workflow (optional, used with workflowId)
     * @param params.triggerName - Specific trigger name within the workflow (optional, used with workflowId)
     * @param params.input - Input data to pass to the workflow
     * @returns The webhook response data
     *
     * @example
     * ```typescript
     * // By workflow ID (recommended)
     * const result = await client.triggers.fireWebhook({
     *   workflowId: 123,
     *   input: { query: 'hello world' }
     * });
     *
     * // By workflow ID + trigger name
     * const result = await client.triggers.fireWebhook({
     *   workflowId: 123,
     *   triggerName: 'My Webhook',
     *   input: { query: 'hello world' }
     * });
     *
     * // By direct URL
     * const result = await client.triggers.fireWebhook({
     *   triggerUrl: 'https://api.openserv.ai/webhooks/trigger/TOKEN',
     *   input: { query: 'hello world' }
     * });
     * ```
     */
    async fireWebhook(params) {
        if (params.triggerUrl) {
            // Direct URL provided -- use raw POST (the URL may point to an external host)
            return this.client.post(params.triggerUrl, params.input || {});
        }
        const token = await this.resolveWebhookToken(params);
        return this.client.post(`/webhooks/trigger/${token}`, params.input || {});
    }
    /**
     * Resolve a webhook trigger token from workflowId.
     */
    async resolveWebhookToken(params) {
        if (!params.workflowId) {
            throw new Error("Either workflowId or triggerUrl is required.");
        }
        const triggers = await this.list({ workflowId: params.workflowId });
        let trigger;
        if (params.triggerId) {
            trigger = triggers.find((t) => t.id === params.triggerId && t.token);
        }
        else if (params.triggerName) {
            trigger = triggers.find((t) => t.name === params.triggerName && t.token);
        }
        else {
            trigger = triggers.find((t) => t.token && !t.props?.x402Pricing);
        }
        if (!trigger?.token) {
            const hint = params.triggerId
                ? `Trigger ${params.triggerId}`
                : params.triggerName
                    ? `Trigger "${params.triggerName}"`
                    : "No webhook trigger";
            throw new Error(`${hint} not found or has no token in workflow ${params.workflowId}.`);
        }
        return trigger.token;
    }
    /**
     * Get callable triggers for a workspace.
     *
     * Returns the list of triggers that can be called externally, along with
     * their input schemas and endpoint URLs. This is used during ERC-8004
     * deployment to register the agent's available services on-chain.
     *
     * @param params - Parameters
     * @param params.workflowId - The workflow (workspace) ID
     * @returns Array of callable triggers with their schemas and endpoints
     *
     * @example
     * ```typescript
     * const triggers = await client.triggers.getCallableTriggers({ workflowId: 123 });
     * for (const trigger of triggers) {
     *   console.log(trigger.name, trigger.webEndpoint);
     * }
     * ```
     */
    async getCallableTriggers(params) {
        return this.client.get(`/workspaces/${params.workflowId}/callable-triggers`);
    }
}
exports.TriggersAPI = TriggersAPI;
