// @openserv-labs/client/dist/x402.js
"use strict";
/**
 * Vendored x402 EVM client-side payment implementation.
 *
 * This replaces the `x402-fetch` and `x402` npm packages to avoid pulling in
 * wagmi, WalletConnect, Reown, MetaMask SDK, Solana SDK, and hundreds of MB
 * of transitive dependencies that are unnecessary for server-side EVM usage.
 *
 * Only EVM payment support is included (Base network), which is all this SDK needs.
 * The x402 protocol specification: https://www.x402.org/
 *
 * @module
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSigner = createSigner;
exports.wrapFetchWithPayment = wrapFetchWithPayment;
const viem_1 = require("viem");
const accounts_1 = require("viem/accounts");
const chains_1 = require("viem/chains");
const node_crypto_1 = require("node:crypto");
const utils_1 = require("./utils");
// ============================================================================
// Constants
// ============================================================================
/** Map network names to viem Chain objects. */
const NETWORK_CHAINS = {
    base: chains_1.base,
    "base-sepolia": chains_1.baseSepolia,
};
/**
 * EIP-712 typed data for USDC TransferWithAuthorization (EIP-3009).
 * Used to sign USDC transfer authorizations without an on-chain transaction.
 */
const TRANSFER_WITH_AUTHORIZATION_TYPES = {
    TransferWithAuthorization: [
        { name: "from", type: "address" },
        { name: "to", type: "address" },
        { name: "value", type: "uint256" },
        { name: "validAfter", type: "uint256" },
        { name: "validBefore", type: "uint256" },
        { name: "nonce", type: "bytes32" },
    ],
};
// ============================================================================
// Helpers
// ============================================================================
function base64Encode(str) {
    return Buffer.from(str).toString("base64");
}
function createNonce() {
    return (0, viem_1.toHex)((0, node_crypto_1.randomBytes)(32));
}
function encodePayment(payment) {
    const safe = {
        ...payment,
        payload: {
            ...payment.payload,
            authorization: Object.fromEntries(Object.entries(payment.payload.authorization).map(([key, value]) => [
                key,
                typeof value === "bigint" ? value.toString() : value,
            ])),
        },
    };
    return base64Encode(JSON.stringify(safe));
}
// ============================================================================
// Payment Logic
// ============================================================================
/**
 * Select the best payment requirement from the server's options.
 * Prefers requirements matching the wallet's network and "exact" scheme.
 */
function selectPaymentRequirements(requirements, network) {
    if (network) {
        const matching = requirements.filter((r) => r.scheme === "exact" && r.network === network);
        if (matching.length > 0)
            return matching[0];
    }
    // Fallback: first "exact" scheme requirement, then any
    const exact = requirements.filter((r) => r.scheme === "exact");
    return exact[0] ?? requirements[0];
}
/**
 * Sign an EIP-3009 TransferWithAuthorization using the wallet client.
 */
async function signAuthorization(walletClient, authorization, paymentRequirements) {
    const chainId = walletClient.chain?.id;
    if (!chainId) {
        throw new Error("Wallet client must have a chain configured");
    }
    const signature = await walletClient.signTypedData({
        account: walletClient.account,
        types: TRANSFER_WITH_AUTHORIZATION_TYPES,
        domain: {
            name: paymentRequirements.extra?.name,
            version: paymentRequirements.extra?.version,
            chainId,
            verifyingContract: (0, viem_1.getAddress)(paymentRequirements.asset),
        },
        primaryType: "TransferWithAuthorization",
        message: {
            from: (0, viem_1.getAddress)(authorization.from),
            to: (0, viem_1.getAddress)(authorization.to),
            value: BigInt(authorization.value),
            validAfter: BigInt(authorization.validAfter),
            validBefore: BigInt(authorization.validBefore),
            nonce: authorization.nonce,
        },
    });
    return signature;
}
/**
 * Create a signed x402 payment header for an EVM payment.
 */
async function createPaymentHeader(walletClient, x402Version, paymentRequirements) {
    const from = walletClient.account.address;
    const nonce = createNonce();
    const validAfter = BigInt(Math.floor(Date.now() / 1000) - 600).toString();
    const validBefore = BigInt(Math.floor(Date.now() / 1000 + paymentRequirements.maxTimeoutSeconds)).toString();
    const unsignedPayment = {
        x402Version,
        scheme: paymentRequirements.scheme,
        network: paymentRequirements.network,
        payload: {
            signature: undefined,
            authorization: {
                from,
                to: paymentRequirements.payTo,
                value: paymentRequirements.maxAmountRequired,
                validAfter,
                validBefore,
                nonce,
            },
        },
    };
    const signature = await signAuthorization(walletClient, unsignedPayment.payload.authorization, paymentRequirements);
    return encodePayment({
        ...unsignedPayment,
        payload: {
            ...unsignedPayment.payload,
            signature,
        },
    });
}
// ============================================================================
// Public API
// ============================================================================
/**
 * Create a viem WalletClient configured for x402 payment signing.
 *
 * @param network - Network name (e.g., "base", "base-sepolia")
 * @param privateKey - Hex-encoded private key
 * @returns Configured WalletClient
 */
function createSigner(network, privateKey) {
    const chain = NETWORK_CHAINS[network];
    if (!chain) {
        throw new Error(`Unsupported network: ${network}. Supported: ${Object.keys(NETWORK_CHAINS).join(", ")}`);
    }
    return (0, viem_1.createWalletClient)({
        account: (0, accounts_1.privateKeyToAccount)((0, utils_1.normalizePrivateKey)(privateKey)),
        chain,
        transport: (0, viem_1.http)(),
    });
}
/**
 * Wrap a fetch function with x402 payment handling.
 *
 * When the wrapped fetch receives a 402 response, it automatically:
 * 1. Parses the payment requirements from the response
 * 2. Signs an EIP-3009 TransferWithAuthorization
 * 3. Retries the request with the signed payment in the X-PAYMENT header
 *
 * The retry calls the raw (unwrapped) fetch directly, so a second 402
 * is returned to the caller as-is without infinite retry loops.
 *
 * @param fetchFn - The fetch function to wrap
 * @param walletClient - Wallet client for signing payments
 * @param maxValue - Maximum payment amount in atomic units (default: 0.1 USDC = 100000)
 * @returns Wrapped fetch function with x402 payment support
 */
function wrapFetchWithPayment(fetchFn, walletClient, maxValue = BigInt(0.1 * 10 ** 6)) {
    return async (input, init) => {
        const response = await fetchFn(input, init);
        if (response.status !== 402) {
            return response;
        }
        // Parse payment requirements from 402 response
        const body = (await response.json());
        // Detect wallet's network name from chain ID
        const chainId = walletClient.chain?.id;
        const network = chainId
            ? Object.entries(NETWORK_CHAINS).find(([, c]) => c.id === chainId)?.[0]
            : undefined;
        const selected = selectPaymentRequirements(body.accepts, network);
        if (BigInt(selected.maxAmountRequired) > maxValue) {
            throw new Error("Payment amount exceeds maximum allowed");
        }
        const paymentHeader = await createPaymentHeader(walletClient, body.x402Version, selected);
        // Retry with payment header using raw fetchFn (not the wrapper) to avoid loops
        return fetchFn(input, {
            ...init,
            headers: {
                ...(init?.headers || {}),
                "X-PAYMENT": paymentHeader,
                "Access-Control-Expose-Headers": "X-PAYMENT-RESPONSE",
            },
        });
    };
}
