// @openserv-labs/client/dist/cli/deploy/api-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiClient = exports.ApiError = void 0;
const axios_1 = __importDefault(require("axios"));
const DEFAULT_ORCHESTRATOR_URL = "https://agent-orchestrator.openserv.ai";
class ApiError extends Error {
    statusCode;
    constructor(message, statusCode) {
        super(message);
        this.statusCode = statusCode;
        this.name = "ApiError";
    }
}
exports.ApiError = ApiError;
class ApiClient {
    client;
    constructor(opts) {
        const headers = {
            "x-openserv-key": opts.apiKey,
        };
        this.client = axios_1.default.create({
            baseURL: opts.orchestratorUrl || DEFAULT_ORCHESTRATOR_URL,
            headers,
            maxBodyLength: 100 * 1024 * 1024,
            maxContentLength: 100 * 1024 * 1024,
        });
    }
    async createContainer() {
        return this.request("POST", "/container/create");
    }
    async getStatus(id) {
        return this.request("GET", `/container/${id}/status`);
    }
    async upload(id, tarBuffer) {
        await this.request("POST", `/container/${id}/upload`, tarBuffer, {
            headers: { "Content-Type": "application/gzip" },
        });
    }
    async exec(id, command, timeout) {
        return this.request("POST", `/container/${id}/exec`, {
            command,
            timeout,
        });
    }
    async start(id, entrypoint) {
        await this.request("POST", `/container/${id}/start`, {
            entrypoint: entrypoint || "npx tsx src/agent.ts",
        });
    }
    async restart(id) {
        await this.request("POST", `/container/${id}/restart`);
    }
    async goLive(id, mode = "on-demand") {
        return this.request("POST", `/container/${id}/go-live`, {
            mode,
        });
    }
    async updateEndpointUrl(agentId, agentApiKey, endpointUrl) {
        const platformUrl = process.env.OPENSERV_API_URL || "https://api.openserv.ai";
        await axios_1.default.put(`${platformUrl}/agents/${agentId}/endpoint-url`, { endpoint_url: endpointUrl }, { headers: { "x-openserv-key": agentApiKey } });
    }
    async request(method, path, body, config) {
        try {
            const res = await this.client.request({
                method,
                url: path,
                data: body,
                ...config,
            });
            return res.data;
        }
        catch (err) {
            const axiosErr = err;
            const statusCode = axiosErr.response?.status;
            const data = axiosErr.response?.data ?? axiosErr.message;
            const detail = typeof data === "string" ? data : JSON.stringify(data);
            throw new ApiError(`${method} ${path} failed (${statusCode ?? "unknown"}): ${detail}`, statusCode);
        }
    }
}
exports.ApiClient = ApiClient;
