// @openserv-labs/client/dist/cli/deploy/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = exports.Spinner = void 0;
exports.elapsed = elapsed;
function supportsColor() {
    if (process.env.NO_COLOR || process.env.TERM === "dumb")
        return false;
    if (process.env.FORCE_COLOR)
        return true;
    return process.stderr.isTTY === true || process.stdout.isTTY === true;
}
const color = supportsColor();
const isTTY = process.stdout.isTTY === true;
const green = color ? "\x1b[38;2;1;254;147;1m" : "";
const red = color ? "\x1b[1;31m" : "";
const yellow = color ? "\x1b[1;33m" : "";
const dim = color ? "\x1b[2m" : "";
const reset = color ? "\x1b[0m" : "";
const clearLine = "\x1b[2K\r";
const frames = ["⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"];
class Spinner {
    msg;
    timer = null;
    i = 0;
    constructor(msg) {
        this.msg = msg;
        if (isTTY) {
            process.stdout.write(`\n${green}  ${frames[0]} ${msg}${reset}`);
            this.timer = setInterval(() => {
                this.i++;
                const frame = frames[this.i % frames.length];
                process.stdout.write(`${clearLine}${green}  ${frame} ${this.msg}${reset}`);
            }, 80);
        }
        else {
            console.log(`\n  => ${msg}`);
        }
    }
    get isSpinning() {
        return this.timer !== null;
    }
    clear() {
        if (this.timer) {
            clearInterval(this.timer);
            this.timer = null;
        }
        if (isTTY) {
            process.stdout.write(clearLine);
        }
    }
    stop(detail) {
        this.clear();
        console.log(`${green}  => ${this.msg}${reset}`);
        if (detail) {
            console.log(`${dim}     ${detail}${reset}`);
        }
    }
    fail(msg) {
        this.clear();
        console.error(`${red}  ✖ ${msg}${reset}`);
    }
    warn(msg) {
        this.clear();
        console.log(`${green}  => ${this.msg}${reset}`);
        console.warn(`${yellow}     ⚠ ${msg}${reset}`);
    }
}
exports.Spinner = Spinner;
exports.logger = {
    step: (msg) => console.log(`\n${green}  => ${msg}${reset}`),
    detail: (msg) => console.log(`${dim}     ${msg}${reset}`),
    success: (msg) => console.log(`\n${green}  ✔ ${msg}${reset}`),
    warn: (msg) => console.warn(`${yellow}     ⚠ ${msg}${reset}`),
    error: (msg) => console.error(`${red}  ✖ ${msg}${reset}`),
    info: (...args) => console.log(...args),
    spin: (msg) => new Spinner(msg),
};
function elapsed(startMs) {
    const sec = (Date.now() - startMs) / 1000;
    return sec < 1 ? `${(sec * 1000).toFixed(0)}ms` : `${sec.toFixed(1)}s`;
}
